runtime/autoload/man.vim

```diff
@@ -68,19 +66,63 @@ function! man#read_page(ref) abort
     " call to s:error() is unnecessary
     return
   endtry
-  call s:read_page(path)
+  let page = s:get_page(path)
+  if empty(page)
+    return
+  endif
+  call s:put_page(page)
 endfunction
 
-function! s:read_page(path) abort
+" Handler for s:system() function.
+function! s:system_handler(jobid, data, event) abort
+  if a:event == 'stdout'
+    let self.output .= join(a:data, "\n")
+  else
+    let s:shell_error = a:data
+  endif
+endfunction
+
+" Run a system command and timeout after 30 seconds.
+function! s:system(cmd, ...) abort
+  let opts = {
+        \ 'output': '',
+        \ 'on_stdout': function('s:system_handler'),
+        \ 'on_exit': function('s:system_handler'),
+        \ }
+  let jobid = jobstart(a:cmd, opts)
+
+  if jobid < 1
+    throw printf('command error %d: %s', jobid, join(a:cmd))
+  endif
+
+  let res = jobwait([jobid], 30000)
+  if res[0] == -1
+    silent! call jobstop(jobid)
+    throw printf('command timed out: %s', join(a:cmd))
+  elseif s:shell_error != 0
+    throw printf("command error (%d) %s: %s", jobid, join(a:cmd), opts.output)
```

Comment:
We should start keeping track of stderr because that is where the error will be, not on stdout. Just rename opts.output to opts.stdout and add opts.stderr and record stderr into there.